src/Reminiscence/Arrays/NativeMemoryArray.cs

```diff
@@ -0,0 +1,141 @@
+﻿#if SUPPORTS_NATIVE_MEMORY_ARRAY
+using System;
+using System.IO;
+using System.Runtime.CompilerServices;
+
+namespace Reminiscence.Arrays
+{
+    /// <summary>
+    /// An array that holds instances of PDS types contiguously in virtual memory.
+    /// </summary>
+    /// <typeparam name="T">
+    /// The type of value to be stored in the array.
+    /// </typeparam>
+    public sealed unsafe class NativeMemoryArray<T> : ArrayBase<T>
+        where T : unmanaged
+    {
+        private readonly IUnmanagedMemoryAllocator allocator;
+
+        private T* head;
+
+        private long length;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="NativeMemoryArray{T}"/> class.
+        /// </summary>
+        /// <param name="allocator">
+        /// The <see cref="IUnmanagedMemoryAllocator"/> to use to allocate, reallocate, and free
+        /// contiguous blocks of virtual memory.
+        /// </param>
+        /// <param name="size">
+        /// The number of elements that the array should be able to store.
+        /// </param>
+        public NativeMemoryArray(IUnmanagedMemoryAllocator allocator, long size)
+        {
+            this.allocator = allocator ?? throw new ArgumentNullException(nameof(allocator));
+            this.Resize(size);
+        }
+
+        /// <summary>
+        /// Finalizes an instance of the <see cref="NativeMemoryArray{T}"/> class.
+        /// </summary>
+        ~NativeMemoryArray() => this.Resize(0);
+
+        /// <inheritdoc />
+        public override T this[long idx]
+        {
+            get => Unsafe.ReadUnaligned<T>(GetPointer(idx));
+            set => Unsafe.WriteUnaligned(GetPointer(idx), value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T* GetPointer(long idx)
+        {
+            // handle non-negative and out-of-bounds with a single test
+            if (unchecked((ulong)idx >= (ulong)this.length))
+            {
+                ThrowArgumentOutOfRangeExceptionForIndex();
+            }
+
+            return this.head + idx;
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static void ThrowArgumentOutOfRangeExceptionForIndex() => throw new ArgumentOutOfRangeException("idx", "Must be non-negative and less than the size of the array.");
+
+        /// <inheritdoc />
+        public override long Length => this.length;
+
+        /// <inheritdoc />
+        public override bool CanResize => true;
+
+        /// <inheritdoc />
+        public override void Dispose()
+        {
+            this.Resize(0);
+            GC.SuppressFinalize(this);
```

Comment:
We should make sure that we can't become "un-disposed" after `Dispose()` is called.

Suggestion is to set a flag in `Dispose()`, and check that flag in `Resize(long)`.  *We only **really** need to check it there, since after `Resize(0)`, everything else is harmless.*